stores/filestore.go

```diff
@@ -3208,6 +3241,148 @@ func (ms *FileMsgStore) lookup(seq uint64) (*pb.MsgProto, error) {
 	return msg, nil
 }
 
+// This is equivalent to readRecord except that we know this is a message record
+// and we know the expected message size so we can do a single read as opposed
+// to read header + read payload.
+func (ms *FileMsgStore) readMsgRecord(r io.Reader, buf []byte, expectedMsgSize uint32) ([]byte, error) {
+	totalSize := recordHeaderSize + int(expectedMsgSize)
+	buf = util.EnsureBufBigEnough(buf, totalSize)
+	if _, err := io.ReadFull(r, buf[:totalSize]); err != nil {
+		return buf, err
+	}
+	header := buf[:recordHeaderSize]
+	msgSize := util.ByteOrder.Uint32(header[:4])
+	if msgSize == 0 {
+		crc := util.ByteOrder.Uint32(header[4:recordHeaderSize])
+		if crc == 0 {
+			return buf, errNeedRewind
+		}
+	}
+	if msgSize != expectedMsgSize {
+		return buf, fmt.Errorf("expected size to be %v, got %v", expectedMsgSize, msgSize)
+	}
+	if ms.fstore.opts.DoCRC {
+		crc := util.ByteOrder.Uint32(header[4:recordHeaderSize])
+		// check CRC against what was stored
+		if c := crc32.Checksum(buf[recordHeaderSize:totalSize], ms.fstore.crcTable); c != crc {
+			return buf, fmt.Errorf("corrupted data, expected crc to be 0x%08x, got 0x%08x", crc, c)
+		}
+	}
+	return buf, nil
+}
+
+// Read from disk and returns the message for the given `seq` but perform
+// a single disk I/O with large buffer to read possibly several subsequent
+// messages. These are added to the cache which will speed up lookups for
+// next message sequence.
+func (ms *FileMsgStore) readAheadMsgs(fslice *fileSlice, seq uint64) (*pb.MsgProto, error) {
+
+	// Compute the offset in the index file itself.
+	idxFileOffset := 4 + (int64(seq-fslice.firstSeq)+int64(fslice.rmCount))*msgIndexRecSize
+	// Then position the file pointer of the index file.
+	if _, err := fslice.idxFile.handle.Seek(idxFileOffset, io.SeekStart); err != nil {
+		return nil, err
+	}
+
+	// Keep track of individual size of messages we are going to bulk read
+	var _msgSizes [2048]int
+	msgSizes := _msgSizes[:0]
+
+	// Keep track of total number of messages and size
+	totalMsgs := 0
+	totalMsgsSize := 0
+	firstMsgOffset := int64(0)
+
+	// Indexes are fixed size. Read a page (4K) worth of them.
+	// Loop if needed if total message size is below readBufSize.
+	idxsBuf := [128 * msgIndexRecSize]byte{}
+	idxSeq := seq
+	maxIdxs := int(fslice.lastSeq - seq + 1)
+	done := false
+	idxBufSize := cap(idxsBuf)
+	for iter := 0; !done; iter++ {
+		if (iter+1)*128 > maxIdxs {
+			idxBufSize = (maxIdxs - (iter * 128)) * msgIndexRecSize
+			// Will be done after reading this page of indexes
+			done = true
+		}
+		b, err := io.ReadFull(fslice.idxFile.handle, idxsBuf[:idxBufSize])
```

Comment:
An index on disk is 32 bytes right now. So in memory it would be 8 bytes for key, 20 for content (since we would not need the CRC - 4 bytes - that is on disk. I could make this another option to allow keeping channel message's index in memory..

Regardless, I wanted to revisit the reading of indexes in this read ahead code (that was not my original approach). Maybe I should use the same buffer size than for messages and read up to that number (or up to the number of indexes in this file slice) instead of reading by pages.

Reply:
I think if we have an active sub, up to some limit, we should keep the whole index for the channel past the floor (all active subs lowest seqno) in memory and avoid file IO.